firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,504 @@
+
+    ..note:: This is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    Shape is a tuple of (mesh.num_vertices(), geometric_dim)."""
+    cdef:
+        PetscInt v, vStart, vEnd, offset
+        PetscInt i, dim = shape[1]
+        np.ndarray[PetscReal, ndim=2, mode="c"] swarm_coords, coords
+
+    # get coords field - NOTE this isn't copied so make sure
+    # swarm.restoreField is called too!
+    swarm_coords = swarm.getField("DMSwarmPIC_coor").reshape(shape)
+    coords = np.empty_like(swarm_coords)
+    vStart = 0
+    vEnd = swarm.getLocalSize()
+
+    for v in range(vStart, vEnd):
+        CHKERR(PetscSectionGetOffset(global_numbering.sec, v, &offset))
+        for i in range(dim):
+            coords[offset, i] = swarm_coords[v - vStart, i]
+
+    # have to restore coords field once accessed to allow access again
+    swarm.restoreField("DMSwarmPIC_coor")
+
+    return coords
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def remove_ghosts_pic(PETSc.DM swarm, PETSc.DM plex):
+    """Remove DMSwarm PICs which are in ghost cells of a distributed
+    DMPlex.
+
+    :arg swarm: The DMSWARM which has been associated with the input
+        DMPlex `plex` using PETSc `DMSwarmSetCellDM`.
+    :arg plex: The DMPlex which is associated with the input DMSWARM
+        `swarm`
+    """
+    cdef:
+        PetscInt cStart, cEnd, ncells, i, npics
+        PETSc.SF sf
+        PetscInt nroots, nleaves
+        const PetscInt *ilocal = NULL
+        const PetscSFNode *iremote = NULL
+        np.ndarray[PetscInt, ndim=1, mode="c"] pic_cell_indices
+        np.ndarray[PetscInt, ndim=1, mode="c"] ghost_cell_indices
+
+    assert plex.handle == swarm.getCellDM().handle
+
+    if plex.comm.size > 1:
+
+        cStart, cEnd = plex.getHeightStratum(0)
+        ncells = cEnd - cStart
+
+        # Get full list of cell indices for particles
+        pic_cell_indices = np.copy(swarm.getField("DMSwarm_cellid"))
+        swarm.restoreField("DMSwarm_cellid")
+        npics = len(pic_cell_indices)
+
+        # Initialise with zeros since these can't be valid ranks or cell ids
+        ghost_cell_indices = np.full(ncells, -1, dtype=IntType)
+
+        # Search for ghost cell indices (spooky!)
+        sf = plex.getPointSF()
+        CHKERR(PetscSFGetGraph(sf.sf, &nroots, &nleaves, &ilocal, &iremote))
+        for i in range(nleaves):
+            if cStart <= ilocal[i] < cEnd:
+                # NOTE need to check this is correct index. Can I check the labels some how?
+                ghost_cell_indices[ilocal[i] - cStart] = ilocal[i]
+
+        # trim -1's to reduce searching needed
+        ghost_cell_indices = ghost_cell_indices[ghost_cell_indices != -1]
+
+        # remove swarm pic parent cell indices which match ghost cell indices
+        for i in range(npics-1, -1, -1):
+            if np.isin(pic_cell_indices[i], ghost_cell_indices):
```

Comment:
O(n^2logn) antipattern. For every call of np.isin, you have to sort `ghost_cell_indices` O(nlogn), the lookup doesn't even use bisection search on the sorted array (so O(n)) and you do this n times.

```
ghost_cell_indices = set(ghost_cell_indices)
for i in range(npics-1, -1, -1)
    if pic_cell_indices[i] in ghost_cell_indices:
        swarm.removePointAtIndex(i)
```

```
In [50]: b = numpy.random.randint(0, 10001, size=20000)                                                                                                                              

In [51]: a = numpy.random.randint(0, 10001, size=200)                                                                                                                                

In [52]: %%timeit 
    ...: for ai in a: 
    ...:     numpy.isin(ai, b) 
    ...:                                                                                                                                                                             
547 ms ± 63 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [53]: %%timeit 
    ...: bset = set(b) 
    ...: for ai in a: 
    ...:     ai in bset 
    ...:                                                                                                                                                                             

6.24 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
I got the sizes the wrong way round, `b` is probably small, but `a` is likely large.

```
In [54]: a = numpy.random.randint(0, 10001, size=20000)                                                                                                                              

In [55]: b = numpy.random.randint(0, 10001, size=200)                                                                                                                                

In [56]:                                                                                                                                                                             

In [56]:                                                                                                                                                                             

In [56]: %%timeit 
    ...: bset = set(b) 
    ...: for ai in a: 
    ...:     ai in bset 
    ...:                                                                                                                                                                             
6.54 ms ± 705 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [57]: %%timeit 
    ...: for ai in a: 
    ...:     numpy.isin(ai, b) 
    ...:                                                                                                                                                                             
2.17 s ± 101 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Yikes. Changed! For future reference, how does making this a `set` first actually help under the hood?

Reply:
sets have O(1) lookup, so you have an O(n) construction cost of turn an array of integers into a set. Then you have n O(1) lookups (1 for each ai in a). So the complexity of each lookup is O(1), you do n of them for O(n), and you have to make a datastructure that costs O(n), so total complexity is O(n).

Rather than numpy.isin, which sorts b O(nlogn), and then does a linear search for ai in the sorted b O(n) for each ai. So the complexity of the isin call is O(nlogn + n) = O(nlogn) and you do n of them, so total complexity is O(n^2 log n).